File: trunk/ver1/area.cpp
#include <cmath>
#include <cstdio>

#include "edge.h"
#include "triangulation.h"

#include "area.h"
#include "geoquant.h"
#include "triposition.h"

Area::Area(Face f, GQIndex& gqi) : GeoQuant(){
  position = new TriPosition(AREA, 1, f.getSerialNumber());
  dataID = AREA;

  Edge ed;
  for(int ii = 0; ii < 3; ii++){
    ed = Triangulation::edgeTable[(*(f.getLocalEdges()))[ii]];
    TriPosition t(LENGTH, 1, ed.getSerialNumber());
    Len[ii] = gqi[ t ];
    Len[ii]->addDependent(this);
  }
}

void Area::recalculate(){
  double l1 = Len[0]->getValue();
  double l2 = Len[1]->getValue();
  double l3 = Len[2]->getValue();

  double s = (l1 + l2 + l3) * 0.5;
  value = sqrt(s * (s - l1) * (s - l2) * (s - l3));
}
   

File: Java/src/geoquant/Area.java
package geoquant;

import java.util.HashMap;

import triangulation.Face;
import triangulation.StdFace;


public class Area extends Geoquant {
  // Index map
  private static HashMap<TriPosition, Area> Index = new HashMap<TriPosition, Area>();

  private Length lij;
  private Length ljk;
  private Length lik;
  
  public Area(Face f) {
    super(f);
    StdFace stdF = new StdFace(f);
    lij = Length.at(stdF.e12);
    ljk = Length.at(stdF.e23);
    lik = Length.at(stdF.e13);
    
    lij.addObserver(this);
    ljk.addObserver(this);
    lik.addObserver(this);
  }
  
  protected void recalculate() {
    double l1 = lij.getValue();
    double l2 = ljk.getValue();
    double l3 = lik.getValue();
    
    double s = (l1 + l2 + l3) * 0.5;
    value = Math.sqrt(s * (s - l1) * (s - l2) * (s - l3));

  }

  public void remove() {
    deleteDependents();
    lij.deleteObserver(this);
    ljk.deleteObserver(this);
    lik.deleteObserver(this);
    Index.remove(pos);
  }

  public static Area at(Face f) {
    TriPosition T = new TriPosition(f.getSerialNumber());
    Area q = Index.get(T);
    if(q == null) {
      q = new Area(f);
      q.pos = T;
      Index.put(T, q);
    }
    return q;
  }
  
  public static Area At(Face f) {
    TriPosition T = new TriPosition(f.getSerialNumber());
    Area q = Index.get(T);
    if(q == null) {
      q = new Area(f);
      q.pos = T;
      Index.put(T, q);
    }
    return q;
  }
  
  public static double valueAt(Face f) {
    return at(f).getValue();
  }
}
